### ISSUE TITLE
Create npc movement/battle demo

### ISSUE THREAD

**Issue by amymorrill** [uchicago-cs/chiventure#1027]
For the final presentation, we need to create a demo of the work we've done this quarter. In particular, we have improved the ability of npcs to move between rooms and to participate in battles. Accordingly, we are thinking of creating a demo in which an npc moves into an "arena" room and then battles the player. We may also have other parts of the demo in which we show other features we've worked on this quarter (such as random movement). We will first design the demo and then move on to implementation.

**Comment by torresszy** [comment]
After a meeting with Amy earlier today, we have come up with the following plan for the final presentation: we will most likely monkeypatch the game with two NPCs. The player will meet the NPCs in a arena room and have battles with both of them, one of them will be killed and the other one will surrender. This demo should showcase our works in both the NPC movement module and the NPC battle module. The current logistics we have is the following:

1. Player and NPC1 and NPC2 start in an initial room
    - a command that prints out all the NPCs in a room and see NPC1 and NPC2 (need CLI for this)
    - Player moves to the arena, the NPCs will move too
2. Prints out all the NPCs again in the arena room to see that our NPCs has moved
    - Check items in the room
    - Player battles NPC1
    - NPC1 will surrender (still alive → we can have a function to check that)
    - Print out “the NPC has surrendered”
3. Player battles NPC2
    - Kills NPC2
    - Checks items again (now there are the NPC2’s items too) 
4. Player goes back to the initial room
    - Check no NPC in this room anymore

**Comment by torresszy** [comment]
Most of our plan for our demo has been successfully implemented in `battle_move_example_2021`. The main features we want to showcase are battling with NPCs and moving NPCs between rooms. In this demo, we will have two NPCs, `hostile_harry` and `friendly_fiona`; we will also have two rooms, `lobby` and `arena`. In this demo, we will have the player will start in the lobby with both NPCs, and it will move to the `arena` room where they can battle the NPCs. After the battle, the player will return to the lobby. Through these actions, we want to show:
- NPC movements from `lobby` to `arena`
- A CLI command that prints out all the NPCs in a given room
- The ability to battle with the NPCs through the `ATTACK` command and loot the NPCs when they are killed.
- NPCs is set to surrender after their health drop below a threshold, in this demo, `friendly_fiona` will surrender after a few attacks.

**Comment by torresszy** [comment]
I have made a few new updates to the demo:
- A few simple dialogues are added to the example and the player is able to talk to both NPCs now. Chiventure already supports talking with NPCs through the `TALK` command, but it only supports single-word `npc_id`, so I changed `friendly_fiona` and `hostile_harry` to `FIONA` and `HARRY`. Now we can type in `talk to fiona` or `talk to harry` in the CLI to start a conversation with them.
- There was a `segfault` bug when calling lobby and arena more than once is solved, now we can move between lobby and arena multiple times without a problem
- I have updated the comments explaining each of our monkey-patched functions
- When printing out a list of NPCs in the room to the CLI, it will check their healths and only prints those whose health is above 0
- The `ATTACK` command is only allowed in the arena room now, since that is set to be the designated battle room.

On the other hand, there seem to be some minor issues with a dead NPC dropping its items and we are in the process of fixing it.
